WebDriverAgentLib/Categories/XCUIElement+FBWebDriverAttributes.m

```diff
@@ -40,125 +40,198 @@ - (id)forwardingTargetForSelector:(SEL)aSelector
 
 @end
 
+static char const * const FBCachedAttributesKey = "FBCachedAttributes";
 
 @implementation XCElementSnapshot (WebDriverAttributes)
 
+@dynamic fb_cachedAttributes;
+
 - (id)fb_valueForWDAttributeName:(NSString *)name
 {
   return [self valueForKey:[FBElementUtils wdAttributeNameForAttributeName:name]];
 }
 
-- (NSString *)wdValue
+- (NSMutableDictionary<NSString *, id> *)fb_cachedAttributes
 {
-  id value = self.value;
-  if (self.elementType == XCUIElementTypeStaticText) {
-    value = FBFirstNonEmptyValue(self.value, self.label);
-  }
-  if (self.elementType == XCUIElementTypeButton) {
-    value = FBFirstNonEmptyValue(self.value, (self.isSelected ? @YES : nil));
-  }
-  if (self.elementType == XCUIElementTypeSwitch) {
-    value = @([self.value boolValue]);
-  }
-  if (self.elementType == XCUIElementTypeTextView ||
-      self.elementType == XCUIElementTypeTextField ||
-      self.elementType == XCUIElementTypeSecureTextField) {
-    value = FBFirstNonEmptyValue(self.value, self.placeholderValue);
+  return (NSMutableDictionary<NSString *, id> *)objc_getAssociatedObject(self, FBCachedAttributesKey);
+}
+
+- (void)setFb_cachedAttributes:(NSMutableDictionary<NSString *, id> *)newCachedAttributes
+{
+  objc_setAssociatedObject(self, FBCachedAttributesKey, newCachedAttributes, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
+}
+
+- (id)fb_cachedValueFor:(NSString *)attributeName valueGetter:(id (^)(void))valueGetter
+{
+  if (nil == self.fb_cachedAttributes) {
+    [self setFb_cachedAttributes:[NSMutableDictionary new]];
   }
-  value = FBTransferEmptyStringToNil(value);
-  if (value) {
-    value = [NSString stringWithFormat:@"%@", value];
+  id result = [self.fb_cachedAttributes objectForKey:attributeName];
+  if (nil != result) {
+    if ([NSNull null] == result) {
+      return nil;
+    }
+    return result;
```

Comment:
1) You never actually fill the cache
2) This caches on snapshot level, which is ok, as resolving element will change snapshot. However I was pretty sure Apple does this anyway?

Reply:
1. Oh, sorry. It's a cleanup issue after I removed perf logging. Now should be better.

2. Aple probably does, but we should as well, since we have additional stuff built on top of it. The perf log I got confirms that.